app/module/drivers/sensor/battery/battery_common.c

```diff
@@ -29,15 +29,37 @@ int battery_channel_get(const struct battery_value *value, enum sensor_channel c
     return 0;
 }
 
-uint8_t lithium_ion_mv_to_pct(int16_t bat_mv) {
-    // Simple linear approximation of a battery based off adafruit's discharge graph:
-    // https://learn.adafruit.com/li-ion-and-lipoly-batteries/voltages
-
-    if (bat_mv >= 4200) {
-        return 100;
-    } else if (bat_mv <= 3450) {
-        return 0;
+uint8_t lithium_ion_mv_to_pct(int16_t batt_mv) {
+    // Lookup table of slope formulas for calculating remaining battery capacity.
+    // The original values used to calculate slopes come from:
+    //
+    //       https://blog.ampow.com/lipo-voltage-chart/
+    //
+    struct lookup_point {
+        int16_t millivolts;
+        int16_t percent;
+    };
+
+    static const struct lookup_point battery_lookup[] = {
+        {.millivolts = 4200, .percent = 100}, {.millivolts = 3870, .percent = 60},
+        {.millivolts = 3690, .percent = 10},  {.millivolts = 3610, .percent = 5},
+        {.millivolts = 3270, .percent = 0},
+    };
+
+    if (batt_mv > battery_lookup[0].millivolts) {
+        return battery_lookup[0].percent;
     }
 
-    return bat_mv * 2 / 15 - 459;
-}
\ No newline at end of file
+    for (int i = 0; i < ARRAY_SIZE(battery_lookup); i++) {
```

Comment:
Either I missed this on the last review, or you changed it since I looked at it. This loop should start at 1 since you're looking at `i` and `i - 1`

Reply:
Yep, somehow I changed it, though the change was definitely not deliberate. Either way I've fixed this.

I've had a poke around in repo but I can not seem to find any tests for sensor code, admittedly C++ is definitely not my best language, but when I used to code for a living I'd typically implement unit tests for something like this (partly for my own sanity). Is there anything I can use as a reference for implementing some tests?

Reply:
We have test for keymap features, but those work by logging output and diffing it against the expected output. I'm not sure how difficult it would be to bring in a unit testing framework for general code.

Reply:
Merged again, I noticed there are a few more changes since last time :)